### PR TITLE
Zanj datasets getitem

### PR DESCRIPTION
**Pull request by valedan**
Main changes:
- `MazeDataset.__getitem__` returns a `SolvedMaze`
  - Tokenization for the training loop is handled in the dataloader
- Move `maze_to_tokens` to `LatticeMaze.as_tokens()`
- Added `MazeDataset.as_tokens` replacing `mazes_tokens`
- Remove `mazes_objs`, `mazes_tokens`, `mazes_array`
- Add some tests for dataloader, finish tests for MazeDataset
- Ran `make format`

### CODE REVIEW

File: maze_transformer/evaluation/eval_model.py
Comment (mivanit) [inline maze_transformer/evaluation/eval_model.py:154]: we spoke today about having the tokenization be done by the model (since that is where we have access to the `ConfigHolder`, but changing that can wait

File: maze_transformer/training/dataset.py
Comment (mivanit) [inline maze_transformer/training/dataset.py:115]: yeah, this class should be deleted. I might add it to muutils though since it seems generally useful.

File: maze_transformer/training/maze_dataset.py
Comment (mivanit) [reply maze_transformer/training/maze_dataset.py:166]: neither required for this PR, certainly not migration of tokenization logic

File: maze_transformer/training/maze_dataset.py
Comment (mivanit) [inline maze_transformer/training/maze_dataset.py:364]: I absolutely did not realize that `np.percentile` took a percent between 0 and 100. maybe add a comment to the docstring about this, seems like a gotcha?

File: maze_transformer/training/training.py
Comment (mivanit) [inline maze_transformer/training/training.py:44]: im not sure if making this a list comprehension (I assume this is what you mean by vectorizing given that I dont think this is numpy-vectorizable) would make this faster, but it would be cleaner

File: maze_transformer/training/training.py
Comment (mivanit) [inline maze_transformer/training/training.py:97]: I don't thinks `loss` is actually a plain tensor here, I'm pretty sure its a `TransformerLens` Loss object?

File: tests/unit/maze_transformer/generation/test_maze_dataset.py
Comment (mivanit) [inline tests/unit/maze_transformer/generation/test_maze_dataset.py:54]: replace this and other instances if we remove `get()`

**Review by mivanit — APPROVED** [review]
LGTM! only minor notes, which can wait until we merge back into `zanj-integration-datasets`

File: maze_transformer/training/training.py
Comment (valedan) [reply maze_transformer/training/training.py:44]: I did mean the numpy/pytorch version, and I realize that it would be a big chunk of work... but it seems maybe possible? The connection_list and solution are already numpy arrays, so it seems possible that would could take a batch of maze objects and turn it into a batch of tokens (the integer kind) using only tensor operations. I could totally be wrong about the feasibility here. This would be TONS of work, and my benchmarking shows that this collation contributes less than 10% to the runtime of the overall training loop, so definitely not worth it right now!

File: maze_transformer/training/dataset.py
Comment (mivanit) [inline maze_transformer/training/dataset.py:411]: what's the reason for removing this comment?

Comment (valedan) [reply maze_transformer/training/dataset.py:411]: The first part:

```    
```python
    dataset_cls._FILTER_NAMESPACE = filter_namespace_cls
    filter_namespace_cls._BASE_DATASET = dataset_cls
```
```

Is just an exact duplicate of the code a couple lines down, it doesn't explain anything.

The second part

`and also adds a `_filter_namespace_cls` attribute to every static method in the namespace class for use in `register_wrap_dataset_filter``

Is no longer true because I removed that part of the method (as suggested in the TODO comment).

File: maze_transformer/training/training.py
Comment (valedan) [reply maze_transformer/training/training.py:97]: Good point! I think we can narrow it down though, in this case we know it is actually a `SingleLoss` because we're not setting `loss_per_token=True` in our forward call. 

These are the relevant TL types:
```
SingleLoss = Float[torch.Tensor, ""] # Type alias for a single element tensor
LossPerToken = Float[torch.Tensor, "batch pos-1"]
Loss = Union[SingleLoss, LossPerToken]
```

Definitely better to use `SingleLoss` here rather than what I've got, I'll make that change. Also open to changing it back to `Loss` if you think that's best.

**Pull request merged**
